deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java

```diff
@@ -213,6 +214,7 @@ public double getL2ByParam(String variable) {
      * Fluent interface for building a list of configurations
      */
     public static class ListBuilder extends MultiLayerConfiguration.Builder {
+        private AtomicInteger layerCounter = new AtomicInteger(-1); //Used only for .layer(Layer) method
```

Comment:
AtomicInteger sounds like ("communicates") you're expecting concurrency, while that is not true - the use of an ordinary map rules out concurrency. You could thus just use an int. 

However, I think it would be wiser to just use the "find max key in map" logic and then just invoke the indexed variant with max+1, as that is very straightforward and immediately understandable for a code-reader, and there is no possibility to start wondering what happens if you mix the methods. Right now, while reviewing this, I am wondering what happens if I e.g. add layers using index in out-of-order fashion, and then add one layer using the non-indexed method. After thinking a bit, I can't immediately see how that would fail - but my point is that this particular thinking should really not be necessary!

Also, JavaDoc, explaining what they do, and how they interact.

I would note that the "same index overrides" has tripped me up over a dozen times already, and I am still just getting to know the API! It should rather have thrown - it makes NO sense that this is possible, IMHO. It should at least be JavaDocced, and could potentially also output a log.warn(..)